anndata/experimental/merge.py

```diff
@@ -0,0 +1,602 @@
+                ds_part = None
+                if axis == 0:
+                    ds_part = ds[idx : idx + slice_size, :]
+                elif axis == 1:
+                    ds_part = ds[:, idx : idx + slice_size]
+
+                yield ri(ds_part, axis=1 - axis, fill_value=fill_value)
+                rem_slices -= slice_size
+                idx += slice_size
+
+
+###################
+# File Management
+###################
+
+
+@singledispatch
+def as_group(store, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    raise NotImplementedError("This is not yet implemented.")
+
+
+@as_group.register
+def _(store: os.PathLike, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    if store.suffix == ".h5ad":
+        import h5py
+
+        return h5py.File(store, *args, **kwargs)
+    import zarr
+
+    return zarr.open_group(store, *args, **kwargs)
+
+
+@as_group.register
+def _(store: str, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    return as_group(Path(store), *args, **kwargs)
+
+
+@as_group.register(ZarrGroup)
+@as_group.register(H5Group)
+def _(store, *args, **kwargs):
+    return store
+
+
+###################
+# Reading
+###################
+
+
+def read_as_backed(group: Union[ZarrGroup, H5Group]):
+    """
+    Read the group until
+    SparseDataset, Array or EAGER_TYPES are encountered.
+    """
+
+    def callback(func, elem_name: str, elem, iospec):
+        if iospec.encoding_type in SPARSE_MATRIX:
+            return SparseDataset(elem)
+        elif iospec.encoding_type in EAGER_TYPES:
+            return read_elem(elem)
+        elif iospec.encoding_type == "array":
+            return elem
+        elif iospec.encoding_type == "dict":
+            return {k: read_as_backed(v) for k, v in elem.items()}
+        else:
+            return func(elem)
+
+    return read_dispatched(group, callback=callback)
+
+
+def _df_index(df: Union[ZarrGroup, H5Group]) -> pd.Index:
+    index_key = df.attrs["_index"]
+    return pd.Index(read_elem(df[index_key]))
+
+
+###################
+# Writing
+###################
+
+
+def write_concat_dense(
+    arrays: Sequence[Union[ZarrArray, H5Array]],
+    output_group: Union[ZarrGroup, H5Group],
+    output_path: Union[ZarrGroup, H5Group],
+    axis: Literal[0, 1] = 0,
+    reindexers: Reindexer = None,
+    fill_value=None,
+):
+    """
+    Writes the concatenation of given dense arrays to disk using dask.
+    """
+    import dask.array as da
+
+    def _gen_elem_to_append(elems):
+        for elem, ri in zip(elems, reindexers):
+            yield ri(elem, axis=1 - axis, fill_value=fill_value)
+
+    darrays = None
+    init_elem = arrays[0]
+
+    if isinstance(init_elem, H5Array):
+        darrays = (da.from_array(a, chunks="auto") for a in arrays)
+    elif isinstance(init_elem, ZarrArray):
+        darrays = (da.from_zarr(a) for a in arrays)
+
+    res = da.concatenate(
+        _gen_elem_to_append(
+            darrays,
+        ),
+        axis=axis,
+    )
```

Comment:
I think it's a bit easier to read inline. I don't love having the reindexers and the fill_values being closed over in a function that's called once.

```suggestion

    darrays = None
    init_elem = arrays[0]

    if isinstance(init_elem, H5Array):
        darrays = (da.from_array(a, chunks="auto") for a in arrays)
    elif isinstance(init_elem, ZarrArray):
        darrays = (da.from_zarr(a) for a in arrays)

    res = da.concatenate(
        [ri(a, axis=1 - axis, fill_value=fill_value) for a, ri in zip(darrays, reindexers)],
        axis=axis,
    )
```